pkg/certificate/mrc_reconciler.go

```diff
@@ -1,214 +1,228 @@
-// TODO(5046): add unit tests for this, and other methods in here.
-func shouldUpdateMRCComponentStatus(mrc *v1alpha2.MeshRootCertificate) bool {
-	if isTerminal(mrc) {
-		return false
+// getSigningAndValidatingMRCs returns the signing and validating MRCs from a list of MRCs
+func getSigningAndValidatingMRCs(mrcList []*v1alpha2.MeshRootCertificate) (*v1alpha2.MeshRootCertificate, *v1alpha2.MeshRootCertificate, error) {
+	if len(mrcList) == 0 {
+		log.Error().Err(ErrNoMRCsFound).Msg("when handling MRC event, found no MRCs in OSM control plane namespace")
+		return nil, nil, ErrNoMRCsFound
+	}
+	if len(mrcList) > 2 {
+		log.Error().Err(ErrNumMRCExceedsMaxSupported).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrNumMRCExceedsMaxSupported)).
+			Msgf("expected 2 or less MRCs in the OSM control plane namespace, found %d", len(mrcList))
+		return nil, nil, ErrNumMRCExceedsMaxSupported
 	}
 
-	if len(mrc.Status.Conditions) == 0 {
-		return false
+	var mrc1, mrc2 *v1alpha2.MeshRootCertificate
+	mrc1 = mrcList[0]
+	if len(mrcList) == 2 {
+		mrc2 = mrcList[1]
+	} else {
+		log.Trace().Msgf("found single MRC in the mesh when handling MRC event for MRC %s", mrc1.Name)
+		// if there is only one MRC, set mrc2 equal to mrc1
+		mrc2 = mrc1
 	}
 
-	transitionAfter := mrc.Status.TransitionAfter
-	if transitionAfter == nil {
-		return false
+	if mrc1 == nil || mrc2 == nil {
+		log.Error().Err(ErrUnexpectedNilMRC).Msg("unexpected nil MRC provided when validating MRC intents")
+		return nil, nil, ErrUnexpectedNilMRC
 	}
 
-	//The truncate is needed so we truncate the monotonic clock.
-	return transitionAfter.Truncate(0).After(time.Now())
-}
+	intent1 := mrc1.Spec.Intent
+	intent2 := mrc2.Spec.Intent
+
+	if err := validateMRCIntent(intent1); err != nil {
+		return nil, nil, err
+	}
+	if err := validateMRCIntent(intent2); err != nil {
+		return nil, nil, err
+	}
 
-// shouldUpdateState (and conditions).
-func shouldUpdateState(mrc *v1alpha2.MeshRootCertificate) bool {
-	// TODO(5046): determine the status to we need to be at to be able to move forward.
-	var status v1alpha2.MeshRootCertificateComponentStatus
-
-	for _, useCase := range allUseCases {
-		switch useCase {
-		case UseCaseValidatingWebhook:
-			if mrc.Status.ComponentStatuses.ValidatingWebhook != status {
-				return false
-			}
-		case UseCaseMutatingWebhook:
-			if mrc.Status.ComponentStatuses.MutatingWebhook != status {
-				return false
-			}
-		case UseCaseXDSControlPlane:
-			if mrc.Status.ComponentStatuses.XDSControlPlane != status {
-				return false
-			}
-		case UseCaseSidecar:
-			if mrc.Status.ComponentStatuses.Sidecar != status {
-				return false
-			}
-		case UseCaseBootstrap:
-			if mrc.Status.ComponentStatuses.Bootstrap != status {
-				return false
-			}
-		case UseCaseGateway:
-			if mrc.Status.ComponentStatuses.Gateway != status {
-				return false
-			}
+	log.Debug().Msgf("validating intent combination of %s and %s", intent1, intent2)
```

Comment:
debug line seems like it should occur before validateMRCIntent call. or consider removing

Reply:
`validateMRCIntent` validates the individual intent. This log line refers to validating the intent combination.